wgpu/src/image/raster.rs

```diff
@@ -132,3 +145,67 @@ impl Cache {
         self.map.contains_key(&handle.id())
     }
 }
+
+#[derive(Debug, Clone, Copy)]
+enum Orientation {
+    Default,
+    FlippedHorizontally,
+    FlippedVertically,
+    Rotated90,
+    Rotated180,
+    Rotated270,
+    Rotated90AndFlippedHorizontally,
+    Rotated90AndFlippedVertically,
+}
+
+impl Orientation {
+    // Meaning of the returned value is described e.g. at:
+    // https://magnushoff.com/articles/jpeg-orientation/
+    fn from_exif<R>(reader: &mut R) -> Result<Self, exif::Error>
+    where
+        R: std::io::BufRead + std::io::Seek,
+    {
+        let exif = exif::Reader::new().read_from_container(reader)?;
+
+        Ok(exif
+            .get_field(::exif::Tag::Orientation, ::exif::In::PRIMARY)
+            .and_then(|field| field.value.get_uint(0))
+            .map(|value| match value {
+                2 => Orientation::FlippedHorizontally,
+                3 => Orientation::Rotated180,
+                4 => Orientation::FlippedVertically,
+                5 => Orientation::Rotated90AndFlippedHorizontally,
+                6 => Orientation::Rotated90,
+                7 => Orientation::Rotated90AndFlippedVertically,
+                8 => Orientation::Rotated270,
+                _ => Orientation::Default,
+            })
+            .unwrap_or(Orientation::Default))
+    }
+
+    fn apply(
+        self,
+        mut img: ::image_rs::ImageBuffer<::image_rs::Bgra<u8>, Vec<u8>>,
+    ) -> image_rs::ImageBuffer<::image_rs::Bgra<u8>, Vec<u8>> {
+        use image_rs::imageops::*;
+
+        match self {
+            Self::FlippedHorizontally => flip_horizontal_in_place(&mut img),
+            Self::Rotated180 => rotate180_in_place(&mut img),
+            Self::FlippedVertically => flip_vertical_in_place(&mut img),
+            Self::Rotated90AndFlippedHorizontally => {
+                img = rotate90(&img);
+                flip_horizontal_in_place(&mut img);
+            }
+            Self::Rotated90 => img = rotate90(&img),
+            Self::Rotated90AndFlippedVertically => {
```

Comment:
[nit] So, in theory, I think one could argue that if *we* need to rotate the image 90°, it technically means *the source image* was transformed in reverse "from the correct orientation" - i.e. in case of l.200, the orientation of the JPEG is "Rotated270" from "what was really seen".

The "double transform" ones would then also become reversed in the order of transforms - l.201 being arguably `FlippedVerticallyAndRotated270`.

Alternatively, if we wanted to make it right, we could also try renaming the "Orientation" enum to something else to better capture that it represents "How To Fix" orientation, not just orientation that the specific JPEG "has". Unfortunately, I don't currently have a great idea for what name could be cool for that.

With that said, especially with this being private, that's *totally* a nitpick, so just saying for your awareness and decision at your discretion.

Reply:
Good point. I have refactored the `Orientation` enum into an `Operation` bitflag in fb97fea438ff803c990e36119977c08b0e5773e1. I believe this also sets everything up for future optimizations (i.e. applying a transform matrix to the image once).

Hopefully I didn't break anything!

Reply:
It looks like flipping along the diagonal isn't what I think it is. Fixing!

Reply:
Ok, I think 65c2c17226ec8472066ea48d733a458cb40fe12c should be it!